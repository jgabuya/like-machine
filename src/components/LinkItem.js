import React, { Component } from "react";
import PropTypes from "prop-types";
import {
    Card,
    CardBody,
    CardTitle,
    CardSubtitle,
    CardText,
    CardFooter,
    Button,
    CardImg
} from "reactstrap";
import { FaHeart, FaRegHeart, FaRegTrashAlt } from "react-icons/fa";

import moment from "moment";
import normalizeUrl from "normalize-url";

import "animate.css";

const propTypes = {
    id: PropTypes.string.isRequired,
    title: PropTypes.string,
    description: PropTypes.string,
    url: PropTypes.string.isRequired,
    imageUrl: PropTypes.string,
    createdAt: PropTypes.string.isRequired,
    liked: PropTypes.bool.isRequired,
    likeCount: PropTypes.number.isRequired,
    owned: PropTypes.bool.isRequired,
    onLike: PropTypes.func.isRequired,
    onUnlike: PropTypes.func.isRequired,
    onDelete: PropTypes.func.isRequired
};

class LinkItem extends Component {
    state = {
        animatedLike: false
    };

    componentWillReceiveProps(nextProps) {
        this.setState({
            animatedLike: nextProps.liked && !this.props.liked
        });
    }

    render() {
        let LikeIcon;
        let likeBtnTitle;
        let likeBtnOnClickHandler;

        if (this.props.liked) {
            LikeIcon = FaHeart;
            likeBtnTitle = "Unlike";
            likeBtnOnClickHandler = this.props.onUnlike;
        } else {
            LikeIcon = FaRegHeart;
            likeBtnTitle = "Like";
            likeBtnOnClickHandler = this.props.onLike;
        }

        return (
            <Card className="mb-5">
                {this.props.imageUrl !== null && (
                    <CardImg
                        top
                        src={this.props.imageUrl}
                        alt={this.props.title || this.props.url}
                        style={{
                            width: "100%",
                            maxHeight: "20vw",
                            objectFit: "cover"
                        }}
                    />
                )}

                <CardBody>
                    <CardTitle>
                        <a target="_blank" href={normalizeUrl(this.props.url)}>
                            {this.props.title || this.props.url}
                        </a>
                    </CardTitle>

                    <CardSubtitle>{this.props.description}</CardSubtitle>

                    <CardText>
                        <small
                            className="text-muted"
                            title={this.props.createdAt}
                        >
                            {moment(this.props.createdAt)
                                .startOf("day")
                                .fromNow()}
                        </small>
                    </CardText>
                </CardBody>

                {this.props.user && (
                    <CardFooter>
                        <Button
                            onClick={likeBtnOnClickHandler}
                            outline
                            color="primary"
                            title={likeBtnTitle}
                        >
                            <LikeIcon
                                className={
                                    this.state.animatedLike
                                        ? "animated bounce"
                                        : ""
                                }
                            />{" "}
                            <small>{this.props.likeCount}</small>
                        </Button>

                        {this.props.owned && (
                            <Button
                                onClick={this.props.onDelete}
                                outline
                                color="danger"
                                className="float-right"
                                title="Delete"
                            >
                                <FaRegTrashAlt />
                            </Button>
                        )}
                    </CardFooter>
                )}
            </Card>
        );
    }
}

LinkItem.propTypes = propTypes;

export default LinkItem;
