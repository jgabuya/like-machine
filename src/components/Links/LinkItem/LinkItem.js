import React from "react";
import PropTypes from "prop-types";
import {
	Card,
	CardBody,
	CardTitle,
	CardSubtitle,
	CardText,
	CardFooter,
	Button,
	CardImg
} from "reactstrap";
import { FaHeart, FaRegHeart, FaRegTrashAlt } from "react-icons/fa";

import moment from "moment";
import normalizeUrl from "normalize-url";

const propTypes = {
	id: PropTypes.string.isRequired,
	title: PropTypes.string,
	description: PropTypes.string,
	url: PropTypes.string.isRequired,
	imageUrl: PropTypes.string,
	createdAt: PropTypes.string.isRequired,
	liked: PropTypes.bool.isRequired,
	likeCount: PropTypes.number.isRequired,
	owned: PropTypes.bool.isRequired,
	onLike: PropTypes.func.isRequired,
	onUnlike: PropTypes.func.isRequired,
	onDelete: PropTypes.func.isRequired
};

const LinkItem = props => {
	let likeIcon;
    let likeBtnTitle;
    let likeBtnOnClickHandler;

	if (props.liked) {
        likeIcon = <FaHeart />;
        likeBtnTitle = 'Unlike';
        likeBtnOnClickHandler = props.onUnlike
	} else {
        likeIcon = <FaRegHeart />;
        likeBtnTitle = "Like";
        likeBtnOnClickHandler = props.onLike;
	}

	return (
		<Card className="mb-5">
			{props.imageUrl !== null && (
				<CardImg
					top
					src={props.imageUrl}
					alt={props.title || props.url}
					style={{
						width: "100%",
						maxHeight: "20vw",
						objectFit: "cover"
					}}
				/>
			)}

			<CardBody>
				<CardTitle>
					<a target="_blank" href={normalizeUrl(props.url)}>
						{props.title || props.url}
					</a>
				</CardTitle>

				<CardSubtitle>{props.description}</CardSubtitle>

				<CardText>
					<small className="text-muted" title={props.createdAt}>
						{moment(props.createdAt)
							.startOf("day")
							.fromNow()}
					</small>
				</CardText>
			</CardBody>

			<CardFooter>
                <Button onClick={likeBtnOnClickHandler} outline color="primary" title={likeBtnTitle}>
					{likeIcon} <small>{props.likeCount}</small>
				</Button>

				{props.owned && (
					<Button
						onClick={props.onDelete}
						outline
						color="danger"
                        className="float-right"
                        title="Delete"
					>
						<FaRegTrashAlt />
					</Button>
				)}
			</CardFooter>
		</Card>
	);
};

LinkItem.propTypes = propTypes;

export default LinkItem;
