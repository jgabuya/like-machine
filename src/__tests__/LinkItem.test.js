import React from "react";
import LinkItem from "../components/LinkItem";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";

describe("LinkItem", () => {
    const props = {
        id: "test",
        title: "test",
        description: "test",
        url: "google.com/",
        imageUrl: "",
        createdAt: new Date('January 1, 1970').toISOString(),
        liked: false,
        likeCount: 0,
        owned: false,
        onLike: jest.fn(),
        onUnlike: jest.fn(),
        onDelete: jest.fn()
    };

    const component = <LinkItem {...props} />;

    const wrapper = shallow(component);

    it("matches snapshot", () => {
        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("renders without crashing", () => {
        expect(wrapper.length).toEqual(1);        
    });
});
