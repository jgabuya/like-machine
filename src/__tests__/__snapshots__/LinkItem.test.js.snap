// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkItem matches snapshot 1`] = `
<div
  className="mb-5 card"
>
  <img
    alt="test"
    className="card-img-top"
    src=""
    style={
      Object {
        "maxHeight": "20vw",
        "objectFit": "cover",
        "width": "100%",
      }
    }
  />
  <div
    className="card-body"
  >
    <h5
      className="card-title"
    >
      <a
        href="google.com/"
        target="_blank"
      >
        test
      </a>
    </h5>
    <h6
      className="card-subtitle"
    >
      test
    </h6>
    <p
      className="card-text"
    >
      <small
        className="text-muted"
        title="1969-12-31T16:00:00.000Z"
      >
        49 years ago
      </small>
    </p>
  </div>
</div>
`;
